generator/altitude_generator.cpp

```diff
@@ -0,0 +1,237 @@
+  {
+    if (id != m_altitudeAvailabilityBuilder.size())
+    {
+      LOG(LERROR, ("There's a gap in feature id order."));
+      return;
+    }
+
+    bool hasAltitude = false;
+    MY_SCOPE_GUARD(removeTmpDir, [&] ()
+    {
+      m_altitudeAvailabilityBuilder.push_back(hasAltitude);
+    });
+
+    if (!routing::IsRoad(feature::TypesHolder(f)))
+      return;
+
+    f.ParseGeometry(FeatureType::BEST_GEOMETRY);
+    size_t const pointsCount = f.GetPointsCount();
+    if (pointsCount == 0)
+      return;
+
+    TAltitudes altitudes;
+    bool valid = true;
+    TAltitude minFeatureAltitude = kInvalidAltitude;
+    for (size_t i = 0; i < pointsCount; ++i)
+    {
+      TAltitude const a = m_altitudeGetter.GetAltitude(f.GetPoint(i));
+      if (a == kInvalidAltitude)
+      {
+        valid = false;
+        return;
+      }
+
+      if (minFeatureAltitude == kInvalidAltitude)
+        minFeatureAltitude = a;
+      else
+        minFeatureAltitude = min(minFeatureAltitude, a);
+
+      altitudes.push_back(a);
+    }
+    if (!valid)
+      return;
+
+    hasAltitude = true;
+    m_featureAltitudes.push_back(make_pair(id, Altitude(move(altitudes))));
+
+    if (m_minAltitude == kInvalidAltitude)
+      m_minAltitude = minFeatureAltitude;
+    else
+      m_minAltitude = min(minFeatureAltitude, m_minAltitude);
+  }
+
+  bool HasAltitudeInfo() const
+  {
+    return !m_featureAltitudes.empty();
+  }
+
+  bool IsFeatureAltitudesSorted()
+  {
+    return is_sorted(m_featureAltitudes.begin(), m_featureAltitudes.end(),
+                     my::LessBy(&Processor::TFeatureAltitude::first));
+  }
+
+private:
+  IAltitudeGetter & m_altitudeGetter;
+  TFeatureAltitudes m_featureAltitudes;
+  succinct::bit_vector_builder m_altitudeAvailabilityBuilder;
+  TAltitude m_minAltitude;
+};
+}  // namespace
+
+static_assert(sizeof(AltitudeHeader) == 16, "Wrong header size of altitude section.");
+
+namespace routing
+{
+void BuildRoadAltitudes(string const & mwmPath, IAltitudeGetter & altitudeGetter)
+{
+  try
+  {
+    // Preparing altitude information.
+    Processor processor(altitudeGetter);
+    feature::ForEachFromDat(mwmPath, processor);
+
+    if (!processor.HasAltitudeInfo())
+    {
+      LOG(LINFO, ("No altitude information for road features of mwm:", mwmPath));
+      return;
+    }
+
+    CHECK(processor.IsFeatureAltitudesSorted(), ());
+
+    FilesContainerW cont(mwmPath, FileWriter::OP_WRITE_EXISTING);
+    FileWriter w = cont.GetWriter(ALTITUDES_FILE_TAG);
+    AltitudeHeader header;
+    header.minAltitude = processor.GetMinAltitude();
+    int64_t const startOffset = w.Pos();
+    header.Serialize(w);
+    {
+      // Altitude availability serialization.
+      coding::FreezeVisitor<Writer> visitor(w);
+      succinct::bit_vector_builder & builder = processor.GetAltitudeAvailabilityBuilder();
+      succinct::rs_bit_vector(&builder).map(visitor);
+    }
+    header.featureTableOffset = w.Pos() - startOffset;
```

Comment:
As I understand w (FileWriter) position (w.Pos()) is number of bytes starting form mwm beginning. So we need to memorize the beginning of mwm section (startOffset) and then to subtract it from w.Pos.